src/db/database.ts

```diff
@@ -149,33 +158,72 @@ export class Database<T extends Schema> {
           mpFieldWidth,
         );
 
+        // calculate the IDF for every token
+        // N = total # of documents
+        // nt = # of documents containing the term
+
         const valueRef = new ReferencedValue(
           { offset: 0n, length: 0 },
           valueBuf,
         );
         const iter = btree.iter(valueRef);
 
+        // for a given document, the number of times this specific token appears
+        let tfMap = new Map<DataPointer, number>();
         while (await iter.next()) {
           const currentKey = iter.getKey();
 
           if (ReferencedValue.compareBytes(currentKey.value, valueBuf) !== 0) {
             break;
           }
           const mp = iter.getPointer();
-          const data = await this.dataFile.get(
-            Number(mp.offset),
-            Number(mp.offset) + mp.length - 1,
-          );
-          table.insert(data);
+
+          const dp: DataPointer = {
+            start: Number(mp.offset),
+            end: Number(mp.offset) + mp.length - 1,
+          };
+
+          tfMap.set(dp, (tfMap.get(dp) ?? 0) + 1);
         }
-      }
 
-      const data = table.top;
+        const n = entries;
+        const nt = tfMap.size;
+        // The inverse document frequency (IDF) involves # of documents (N) and the # of documents containing the token term
+        const idf = Math.log(1 + (n - nt + 0.5) / nt + 0.5);
+
+        // a parameter used to limit how much a single query term can affect the score for document D.
+        const K1 = 1.2;
+
+        // a parameter used to control the effect of the field length of field t compared to the average field length.
+        const B = 0.75;
+        for (const [key, tf] of tfMap.entries()) {
+          const { start, end } = key;
+          const data = await this.dataFile.get(start, end);
+          const dataRecord = JSON.parse(data);
+          const valueLength = dataRecord[fieldName].length;
+
+          const num = tf * (K1 + 1);
+          const den =
+            tf +
+            K1 *
+              (1 -
+                B +
+                (B * valueLength) /
+                  (valueLength / (totalFieldValueLength / entries)));
```

Comment:
Is this what prettier formats? I would consider breaking this up into separate variables to represent what each of the terms mean, or match the variable names in the formula exactly.